proposals/nnnn-Use-Taskmaster-To-Handle-Queuing-Operations-In-Managers.md

```diff
@@ -0,0 +1,80 @@
+# Use Taskmaster To Handle Queuing Operations in Managers
+
+* Proposal: [SDL-NNNN](nnnn-Use-Taskmaster-To-Handle-Queuing-Operations-In-Managers.md)
+* Author: [Joey Grover](https://github.com/joeygrover), [Bilal Alsharifi](https://github.com/bilal-alsharifi)
+* Status: **Awaiting review**
+* Impacted Platforms: [Java Suite]
+
+## Introduction
+
+This proposal is to include a new third party dependency called [`TaskMaster`](https://github.com/livio/Taskmaster) and start using it in the SDL Java Suite library. [`TaskMaster`](https://github.com/livio/Taskmaster) provides a queue oriented thread management solution to improve how managers handle queueing operations and make their logic easier to maintain and more efficient. This will allow the Java Suite library to better align with the iOS library that natively has this functionality through the OS.
+
+## Motivation
+
+The SDL Java Suite library currently uses `ThreadPoolExecutor` in `ChoiceSetManager` to manage queuing operations and executing them sequentially using one or multiple threads. The `ThreadPoolExecutor` solution is an improvement over the manual way of handling queuing using flags and endless callbacks. However, the `ThreadPoolExecutor` cannot be easily extended to be used in conjunction with other managers simultaneously. Doing so will require each manager to have its own `ThreadPoolExecutor` instance which means each manager will require at least one thread of its own. This may be okay if we have one or two managers that require the ability to queue tasks. However, as the number of managers in the library keeps growing as well as the need for better thread management, the library should have a more scalable solution. This proposal intends to include a third party dependency called `TaskMaster` created by the SDL project maintainer to have a central management for queues and threads for all managers in the SDL Java Suite library. This will allow for a single point of thread management for all the SDL managers.
+
+## Proposed solution
+
+The proposal suggests using [`TaskMaster`](https://github.com/livio/Taskmaster) in SDL Java Suite to create serial queues of operations that need to be executed in order. `ChoiceSetManager`, currently, uses `ThreadPoolExecutor` to queue operations and run them in order. For example, it uses a queue to submit `CheckChoiceVROptionalOperation` then `PresentChoiceSetOperation` to make sure that manager doesn't start presenting choices before knowing whether VR is optional or not on the target head unit. This queuing approach is much better than manually tracking if there is something queued (as the library currently still does  that in `SoftButtonManager` for example). However, the `ThreadPoolExecutor` solution falls short when we need to extend that solution to more managers because each manager and potentially submanagers will be forced to create new threads.
```

Comment:
```suggestion
The proposal suggests using [`TaskMaster`](https://github.com/livio/Taskmaster) in SDL Java Suite to create serial queues of operations that need to be executed in order. `ChoiceSetManager`, currently, uses `ThreadPoolExecutor` to queue operations and run them in order. For example, it uses a queue to submit `CheckChoiceVROptionalOperation` then `PresentChoiceSetOperation` to make sure that manager doesn't start presenting choices before knowing whether VR is optional or not on the target head unit. This queuing approach is much better than manually tracking if there is something queued (as the library currently still does that in `SoftButtonManager` for example). However, the `ThreadPoolExecutor` solution falls short when we need to extend that solution to more managers because each manager and potentially submanagers will be forced to create new threads.
```